WAF: Add WorDBless Support

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR adds `automattic/wordbless` to the WAF package, which will allow us to write integration tests.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a development dependency for `automattic/wordbless`, and related necessary setup in `composer.json` such as the `post-install-cmd` and `post-update-cmd` scripts, and allowing the `roots/wordpress-core-installer` plugin.
* Splits the existing package tests into two separate phpunit projects: `tests/php/unit` and `tests/php/integration`. Each directory has a unique `phpunit.xml.dist` and `bootstrap.php` file, which allows us to run integration tests inside the pseudo-wordpress context, without affecting/breaking the unit tests.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

peb6dq-Dw-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Review the code
* Run the tests

nateweller committed:
Add automattic/wordbless to the waf package

nateweller committed:
changelog

nateweller committed:
Add line break to end of .gitignore

nateweller committed:
Update composer.json

nateweller committed:
Split WAF php tests into two separate test projects

nateweller committed:
Merge branch 'trunk' into add/waf-wordbless

nateweller committed:
Fix missing use statement for Waf_Runner

nateweller committed:
Fixup project versions

nateweller committed:
changelog

ArSn commented on projects/packages/waf/tests/php/integration/test-waf-activation.php:
```diff
@@ -0,0 +1,36 @@
+	 *
+	 * @before
+	 */
+	protected function before() {
```
any reason to not just use `setUp()`?

nateweller replied:
I was running into a conflict between PHPUnit wanting me to declare the return type for `setUp()`, but PHPCS not wanting the return type declared for PHP compatibility.

So it was either

```
protected function setUp() { ... }

> Fatal error: Declaration of WafActivationTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void 
> in /Users/nate/Documents/Code/jetpack/projects/packages/waf/tests/php/integration/test-waf-activation.php on line 21
```
or
```
protect function setUp(): void { ... }

> void return type is not present in PHP version 7.0 or earlier
> (PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations.voidFound)
```

anomiex replied:
Since you're using WorDBless master here, you should probably be using the `WorDBless\BaseTestCase` class it provides. That in turn uses [Yoast's polyfill class](https://github.com/Yoast/PHPUnit-Polyfills#option-1-yoastphpunitpolyfillstestcasestestcase) that uses `set_up()` and other snake-cased names to avoid that problem.

In your unit tests where you're still using `PHPUnit\Framework\TestCase`, I'd recommend following the same convention (although there you'll still need the `@before` directive). Or you could use Yoast's polyfill class if you want to skip the `@before` there too.

nateweller replied:
Thanks for this info @anomiex! Updated to use the `WorDBless\BaseTestCase` and a `set_up()` method tagged with `@before` via [d53f5ca](https://github.com/Automattic/jetpack/pull/28980/commits/d53f5ca990f8ffb1a6a3008d9062b33d3b98861b).

ArSn reviewed (dismissed):
Looks wonderful!

nateweller committed:
Remove unrelated updates to videopress release cycle init

nateweller commented:
Removed an unrelated release cycle initialization for the videopress plugin that resulted from running the project versions fixup script.

@ArSn Can you re-approve? Sorry about that!

anomiex reviewed (commented):
I note that this change makes your `composer test-coverage-html` script not work. To fix that, you'll need to update the command itself, and probably also the `<coverage>` blocks in the `phpunit.xml.dist` files to point to `../../src`.

(No need to worry about the `test-coverage` script, see p1676489265714399-slack-CBG1CP4EN and #28982)

nateweller committed:
Fix test-coverage-html

nateweller committed:
Use WorDBless\BaseTestCase for WAF activation tests class base

nateweller commented:
> I note that this change makes your `composer test-coverage-html` script not work. To fix that, you'll need to update the command itself, and probably also the `<coverage>` blocks in the `phpunit.xml.dist` files to point to `../../src`.

Thanks for catching this @anomiex 🙇‍♂️

I've updated the `test-coverage-html` script to run for both `unit` and `integration` test directories, and fixed the "src" path to point to `../../../src` in [e5770b4](https://github.com/Automattic/jetpack/pull/28980/commits/e5770b4911e8aef2a3952d1b0ba6f96c34057605).

ArSn commented on projects/packages/waf/tests/php/integration/test-waf-activation.php:
```diff
@@ -0,0 +1,36 @@
+	/**
+	 * Test setup.
+	 *
+	 * @before
```
I gather from the previous conversation here that we don't actually need the `@before` on this side due to the polyfill from the `WorDBless\BaseTestCase`? Unless I just understood that wrong it might be better to remove this here, it probably doesn't hurt in this instance either, though.

nateweller replied:
Ah, I misunderstood - removed the `@before` tag here: https://github.com/Automattic/jetpack/pull/28980/commits/5b1acad446ac5a30d14347561f2db80cbd3b699f

Thanks for catching that 👍 👍

nateweller replied:
Whoops - discovered while writing additional tests that the `@before` tag is necessary 😅

Looks like `WorDBless\BaseTestCase` extends `PHPUnit\Framework\TestCase`, and doesn't have its own implementation of `set_up`.

I reverted the commit.

ArSn replied:
oops, ok :)

anomiex replied:
Are you sure about that? https://github.com/Automattic/wordbless/blob/84d9fe02cc3b222d03dd62ad3145a5265beccb0b/src/BaseTestCase.php#L5-L7 shows it using the Yoast base class.

That hasn't been included in a release version yet, but your composer.json is specifying `@dev` so it should be getting the linked version.

nateweller replied:
Interesting, we do have the `@dev` version specified but it does look like the latest `0.4.1` release gets downloaded by composer into `jetpack-waf/vendor/automattic/wordbless`.

nateweller committed:
Remove @before tag from set_up method

nateweller committed:
[not verified] Revert "Remove @before tag from set_up method"

This reverts commit 5b1acad446ac5a30d14347561f2db80cbd3b699f.

anomiex reviewed (commented):
Needs some lock file updates, but otherwise LGTM.

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into add/waf-wordbless

kraftbj committed:
Composer lock update
